compiler/rustc_typeck/src/check/check.rs

```diff
@@ -208,8 +208,8 @@ pub(super) fn check_fn<'a, 'tcx>(
         // case that a newcomer might make, returning a bare trait, and in that case we populate
         // the tail expression's type so that the suggestion will be correct, but ignore all other
         // possible cases.
-        fcx.check_expr(&body.value);
         fcx.require_type_is_sized(declared_ret_ty, decl.output.span(), traits::SizedReturnType);
+        fcx.check_expr(&body.value);
```

Comment:
@estebank I see you added this branch in https://github.com/rust-lang/rust/commit/e53625706106e0227656ddd2fa4d7df54ae2b90e including the FIXME. Did you add the FIXME because it would be great to unify the `ty::Dynamic` case with all the others and remove the check or for another reason? Just trying to figure out what we'd like to have fixed. 🙂

Reply:
@wesleywiser correct, before my change this was branchless code. I want it to return to being branchless code while also supporting the improved diagnostics.